File: util/src/main/java/com/zaze/utils/DisplayUtil.java
package com.zaze.utils;

import android.content.Context;
import android.util.DisplayMetrics;
import android.util.TypedValue;

import com.zaze.utils.log.ZLog;
import com.zaze.utils.log.ZTag;

/**
 * Description : from in.srain.cube.util
 * date : 2016-03-28 - 15:44
 */
public class DisplayUtil {

    private static int screenWidthPixels;
    private static int screenHeightPixels;
    // --------------------------------------------------
    private static float screenDensity;
    private static int screenDensityDpi;
    // --------------------------------------------------
    private static float screenWidthDp;
    private static float screenHeightDp;
    // --------------------------------------------------

    private static DisplayMetrics metrics;
    private static final String TIP_TO_INIT = "请调用init()方法初始化后, 使用该方法！";

    public static void init(Context context) {
        metrics = context.getResources().getDisplayMetrics();
        screenWidthPixels = metrics.widthPixels;
        screenHeightPixels = metrics.heightPixels;
        screenDensity = metrics.density;
        screenDensityDpi = metrics.densityDpi;
        screenWidthDp = dpiFromPx(screenWidthPixels);
        screenHeightDp = dpiFromPx(screenHeightPixels);
    }

    public static DisplayMetrics getMetrics() {
        return metrics;
    }

    public static int getScreenWidthPixels() {
        return screenWidthPixels;
    }

    public static int getScreenHeightPixels() {
        return screenHeightPixels;
    }

    public static float getScreenDensity() {
        return screenDensity;
    }

    public static int getScreenDensityDpi() {
        return screenDensityDpi;
    }

    public static float getScreenWidthDp() {
        return screenWidthDp;
    }

    public static float getScreenHeightDp() {
        return screenHeightDp;
    }

    // --------------------------------------------------
    // --------------------------------------------------

    /**
     * @param dp dp
     * @return px
     * @deprecated {@link DisplayUtil#pxFromDp(float dp)}
     */
    @Deprecated
    public static float dp2px(float dp) {
        return dp * screenDensity;
    }

    // --------------------------------------------------

    /**
     * px 转 dpi
     *
     * @param px
     * @return
     */
    public static float dpiFromPx(int px) {
        if (metrics != null) {
            float densityRatio = metrics.densityDpi * 1.0f / DisplayMetrics.DENSITY_DEFAULT;
            return (px / densityRatio);
        } else {
            ZLog.e(ZTag.TAG_ERROR, TIP_TO_INIT);
            return 0;
        }
    }

    /**
     * dp转px
     *
     * @param dp
     * @return
     */
    public static float pxFromDp(float dp) {
        if (metrics != null) {
            return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, metrics);
        } else {
            ZLog.e(ZTag.TAG_ERROR, TIP_TO_INIT);
            return 0;
        }
    }

    /**
     * sp 转 px
     *
     * @param sp sp
     * @return
     */
    public static float pxFromSp(float sp) {
        if (metrics != null) {
            return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, sp, metrics);
        } else {
            ZLog.e(ZTag.TAG_ERROR, TIP_TO_INIT);
            return 0;
        }
    }

    // --------------------------------------------------


    public static String getDensityDpiName() {
        if (screenDensityDpi <= DisplayMetrics.DENSITY_LOW) {
            return "ldpi";
        } else if (screenDensityDpi <= DisplayMetrics.DENSITY_MEDIUM) {
            return "mdpi";
        } else if (screenDensityDpi <= DisplayMetrics.DENSITY_TV) {
            return "tv";
        } else if (screenDensityDpi <= DisplayMetrics.DENSITY_HIGH) {
            return "hdpi";
        } else if (screenDensityDpi <= DisplayMetrics.DENSITY_XHIGH) {
            return "xhdpi";
        } else if (screenDensityDpi <= DisplayMetrics.DENSITY_XXHIGH) {
            return "xxhdpi";
        } else if (screenDensityDpi <= DisplayMetrics.DENSITY_XXXHIGH) {
            return "xxxhdpi";
        } else {
            return "x????dpi";
        }
    }

//    public static int designedDP2px(float designedDp) {
//        if (screenWidthDp != 320) {
//            designedDp = designedDp * screenWidthDp / 320f;
//        }
//        return dp2px(designedDp);
//    }
//
//    public static void setPadding(final View view, float left, float top, float right, float bottom) {
//        view.setPadding(designedDP2px(left), dp2px(top), designedDP2px(right), dp2px(bottom));
//    }
}

File: util/src/main/java/com/zaze/utils/DisplayUtil.kt
package com.zaze.utils

import android.app.Application
import android.content.Context
import android.content.res.Resources
import android.graphics.Point
import android.os.Build
import android.util.DisplayMetrics
import android.util.TypedValue
import android.view.WindowManager

import com.zaze.utils.log.ZLog
import com.zaze.utils.log.ZTag

/**
 * Description : from in.srain.cube.util
 * date : 2016-03-28 - 15:44
 */
object DisplayUtil {
    private lateinit var displayProfile: DisplayProfile
    private var matchedDisplayProfile: DisplayProfile? = null

    // --------------------------------------------------
    // --------------------------------------------------
    @JvmStatic
    @JvmOverloads
    fun init(application: Application, baseWidthPixels: Int = -1) {
        displayProfile = DisplayProfile(application.resources.displayMetrics)
        val metrics = DisplayMetrics()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            (application.getSystemService(Context.WINDOW_SERVICE) as WindowManager).defaultDisplay.getRealMetrics(
                metrics
            )
        } else {
            metrics.setTo(application.resources.displayMetrics)
        }
        displayProfile.updateRealMetrics(metrics)
        if (baseWidthPixels > 0) {
            metrics.density = (metrics.widthPixels / baseWidthPixels).toFloat()
            metrics.scaledDensity = metrics.density
            metrics.densityDpi = (DisplayMetrics.DENSITY_DEFAULT * metrics.density).toInt()
            matchedDisplayProfile = DisplayProfile(metrics)
        }
        ZLog.i(ZTag.TAG_DEBUG, "displayProfile : $displayProfile")
    }

    @JvmStatic
    fun replaceResource(resources: Resources): Resources {
        matchedDisplayProfile?.let {
            resources.displayMetrics.setTo(it.metrics)
        }
        return resources
    }

    @JvmStatic
    fun getMetrics(): DisplayMetrics {
        return getDisplayProfile().metrics
    }

    @JvmStatic
    fun getScreenWidthDp(): Float {
        return getDisplayProfile().widthDp
    }

    @JvmStatic
    fun getScreenHeightDp(): Float {
        return getDisplayProfile().heightDp
    }

    fun getDisplayProfile(): DisplayProfile {
        return matchedDisplayProfile ?: displayProfile
    }

    // --------------------------------------------------
    @JvmStatic
    @JvmOverloads
    fun getDensityDpiName(metrics: DisplayMetrics = getMetrics()): String {
        val densityDpi = metrics.densityDpi
        return when {
            densityDpi <= DisplayMetrics.DENSITY_LOW -> "ldpi"
            densityDpi <= DisplayMetrics.DENSITY_MEDIUM -> "mdpi"
            densityDpi <= DisplayMetrics.DENSITY_TV -> "tv"
            densityDpi <= DisplayMetrics.DENSITY_HIGH -> "hdpi"
            densityDpi <= DisplayMetrics.DENSITY_XHIGH -> "xhdpi"
            densityDpi <= DisplayMetrics.DENSITY_XXHIGH -> "xxhdpi"
            densityDpi <= DisplayMetrics.DENSITY_XXXHIGH -> "xxxdpi"
            else -> "????dpi"
        }
    }


    // --------------------------------------------------
    // --------------------------------------------------
    /**
     * @param dp dp
     * @return px
     *
     */
    @Deprecated("{@link DisplayUtil#pxFromDp(float dp)}")
    @JvmStatic
    @JvmOverloads
    fun dp2px(dp: Float, metrics: DisplayMetrics = getMetrics()): Float {
        return dp * metrics.density
    }

    // --------------------------------------------------

    /**
     * px 转 dpi
     *
     * @param px
     * @return
     */
    @JvmStatic
    @JvmOverloads
    fun dpiFromPx(px: Int, metrics: DisplayMetrics = getMetrics()): Float {
        return px / metrics.density
    }

    /**
     * dp转px
     *
     * @param dp
     * @return
     */
    @JvmStatic
    @JvmOverloads
    fun pxFromDp(dp: Float, metrics: DisplayMetrics = getMetrics()): Float {
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, metrics)
    }

    /**
     * sp 转 px
     *
     * @param sp sp
     * @return
     */
    @JvmStatic
    @JvmOverloads
    fun pxFromSp(sp: Float, metrics: DisplayMetrics = getMetrics()): Float {
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, sp, metrics)
    }

    data class DisplayProfile(val metrics: DisplayMetrics) {
        val widthPixels: Int
        val heightPixels: Int
        val widthDp: Float
        val heightDp: Float
        val density: Float
        val densityDpi: Int
        var realWidthPixels: Int = 0
        var realHeightPixels: Int = 0

        init {
            widthPixels = metrics.widthPixels
            heightPixels = metrics.heightPixels
            updateRealMetrics(metrics)
            density = metrics.density
            densityDpi = metrics.densityDpi
            widthDp = dpiFromPx(widthPixels, metrics)
            heightDp = dpiFromPx(heightPixels, metrics)
            // ------------------------------------------------------
        }

        fun updateRealMetrics(realMetrics: DisplayMetrics) {
            realWidthPixels = realMetrics.widthPixels
            realHeightPixels = realMetrics.heightPixels
        }

        fun getDensityDpiName(): String {
            return getDensityDpiName(metrics)
        }

        override fun toString(): String {
            return "DisplayProfile(metrics=$metrics, widthPixels=$widthPixels, heightPixels=$heightPixels, widthDp=$widthDp, heightDp=$heightDp, density=$density, densityDpi=$densityDpi, realWidthPixels=$realWidthPixels, realHeightPixels=$realHeightPixels)"
        }
    }
}
